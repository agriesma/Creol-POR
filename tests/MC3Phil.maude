load modelchecker .

mod PROGRAM is
  pr CREOL-MODEL-CHECKER .

op init : -> Configuration .

eq init =

< 'Butler : Cl | Inh: noInh, Par: noAid,
  Att: ('p1 |-> null), ('p2 |-> null), ('p3 |-> null),
  Mtds:
    < 'init : Mtdname | Param: noAid, Latt: noSubst,
		Code: ('p1 ::= new 'Philosopher('this)) ;
		      ('p2 ::= new 'Philosopher('this)) ;
		      ('p3 ::= new 'Philosopher('this)) ; return (emp) > *

    < 'run : Mtdname | Param: noAid, Latt: noSubst, Code: return (emp) > *

    < 'getNeighbor : Mtdname | Param: noAid,
      Latt: ('label |-> null), ('caller |-> null), ('n |-> null),
      Code:
       (if ('equal[[ 'caller # 'p1 ]]) th ('n ::= 'p2)
        el (if ('equal[[ 'caller # 'p2 ]]) th ('n ::= 'p3)
        el (if ('equal[[ 'caller # 'p3 ]]) th ('n ::= 'p1)
        fi) fi) fi) ; (return ( 'n )) >, Ocnt: 0 >

< 'Philosopher : Cl | Inh: noInh, Par: 'butler,
   Att: ('butler |-> null), ('hungry |-> null), 
        ('chopstick |-> null), ('neighbor |-> null),
   Mtds:
   < 'init : Mtdname | Param: noAid, Latt: ('s |-> null), Code:
		   ('chopstick ::= bool(true)) ; ('hungry ::= bool(false)) ; 
         	   ('s ! 'butler . 'getNeighbor(emp)) ; ('s ? ( 'neighbor ) ) ;
         	   return (emp) > *

     < 'think : Mtdname | Param: noAid, Latt: ('label |-> null), ('caller |-> null), ('l |-> null), Code:
            (await ('not[[ 'hungry ]])) ; ('hungry ::= bool(true));
		  (tailcall 'think(emp)) > *

     < 'eat : Mtdname | Param: noAid, Latt: ('label |-> null), ('caller |-> null), ('l |-> null), Code:
            (await 'hungry) ; ('l ! 'neighbor . 'borrowStick(emp)) ; 
                  ('l ? (noAid)) ; (await 'chopstick) ;
                  ('hungry ::= bool(false)) ;
                  ('l ! 'neighbor . 'returnStick(emp)) ; ('l ? (noAid)) ;
		  (tailcall 'eat(emp)) > *

     < 'borrowStick : Mtdname | Param: noAid, Latt: ('label |-> null), ('caller |-> null), Code:
            (await 'chopstick) ; ('chopstick ::= bool(false)) ; 
                   return(emp) > *

     < 'returnStick : Mtdname | Param: noAid, Latt: ('label |-> null), ('caller |-> null), ('l |-> null ), Code:
            ('chopstick ::= bool(true)); return (emp) > *

     < 'run : Mtdname | Param: noAid, Latt: ('label |-> null), ('caller |-> null),
	    ('l |-> null), ('m |-> null), ('n |-> null), Code:
            (('l ! 'this . 'think(emp)) ; free('l)) |||
		  (('m ! 'this . 'eat(emp)) ; free('m)) ; return (emp) >,
      Ocnt: 0 >

    main('Butler, emp)
    .

endm

mod CREOL-PHIL-CHECK is
  pr MODEL-CHECKER .
  pr PROGRAM .
  pr CREOL-PREDICATES .
endm

set show breakdown on .

red modelCheck(init, <> objcnt('Butler, 1)) .

*** red modelCheck(init, [] objcnt('Butler, 1)) .

red modelCheck(init, [] maxobjcnt('Philosopher, 6)) .

*** red modelCheck(init, <> [] objcnt('Butler, 2)) .

quit .
