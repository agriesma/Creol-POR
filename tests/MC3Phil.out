==========================================
reduce in PROGRAM-CHECKER : init .
result Configuration: 
< "Butler" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null,
    "p3" |-> null, Mtds:
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "neighbor" |-> null, Code: if
    "="("caller" # "p1") th "neighbor" ::= "p2" el if "="("caller" # "p2") th
    "neighbor" ::= "p3" el "neighbor" ::= "p1" fi fi ; return("neighbor") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Philosopher"("this") ; "p2" ::= new "Philosopher"("this") ; "p3" ::= new
    "Philosopher"("this") ; return(emp) >,
  Ocnt: 0 >

< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null,
    "chopstick" |-> null, "hungry" |-> null, "neighbor" |-> null, Mtds:
  < "borrowStick" : Mtdname | Param: noVid, Latt: noSubst, Code: await
    "chopstick" ; "chopstick" ::= bool(false) ; return(emp) > *
  < "eat" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, "label:2" |-> null, Code: await "hungry" ; "label:0" ! "neighbor" .
    "borrowStick"(emp) ; await ("label:0" ??) ; "label:0" ?(noVid) ; await
    "chopstick" ; "hungry" ::= bool(false) ; "label:1" ! "neighbor" .
    "returnStick"(emp) ; free("label:1") ; "label:1" ::= null ; "label:2" !
    "this" . "eat"(emp) ; free("label:2") ; "label:2" ::= null ; return(emp) >
    *
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code:
    "chopstick" ::= bool(true) ; "hungry" ::= bool(false) ; "label:0" !
    "butler" . "getNeighbor"(emp) ; "label:0" ?("neighbor") ; return(emp) > *
  < "returnStick" : Mtdname | Param: noVid, Latt: noSubst, Code: "chopstick"
    ::= bool(true) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "think"(emp) ; free("label:0") ; "label:0"
    ::= null ; "label:1" ! "this" . "eat"(emp) ; free("label:1") ; "label:1"
    ::= null ; return(emp) > *
  < "think" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code: await
    "~"("hungry") ; release ; "hungry" ::= bool(true) ; "label:0" ! "this" .
    "think"(emp) ; free("label:0") ; "label:0" ::= null ; return(emp) >,
  Ocnt: 0 >
Bye.
